modules/objdetect/test/test_qrcode.cpp

```diff
@@ -1,74 +1,98 @@
+// of this distribution and at http://opencv.org/license.html.
 
 #include "test_precomp.hpp"
 
-namespace opencv_test { namespace {
 
-TEST(Objdetect_QRCode, regression)
+namespace opencv_test
+{
+
+String qrcode_images_name[] = {
+    "20110817_030.jpg",
+    "20110817_048.jpg",
+    "img_20120226_161648.jpg",
+    "img_2714.jpg",
+    "img_2716.jpg",
+    "img_3011.jpg",
+    "img_3029.jpg",
+    "img_3070.jpg",
+    "qr_test_030.jpg"
+};
+
+// #define UPDATE_QRCODE_TEST_DATA
+#ifdef  UPDATE_QRCODE_TEST_DATA
+
+TEST(Objdetect_QRCode, generate_test_data)
+{
+    String root = cvtest::TS::ptr()->get_data_path() + "qrcode/";
+    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
+    FileStorage file_config(dataset_config, FileStorage::WRITE);
+
+    file_config << "test_images" << "[";
+    size_t images_count = sizeof(qrcode_images_name) / sizeof(String);
+    for (size_t i = 0; i < images_count; i++)
+    {
+        file_config << "{:" << "image_name" << qrcode_images_name[i];
+        String image_path = root + qrcode_images_name[i];
+        std::vector<Point> transform;
+        Mat src = imread(image_path, IMREAD_GRAYSCALE);
+        EXPECT_TRUE(detectQRCode(src, transform));
+        file_config << "x" << "[:";
+        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].x; }
+        file_config << "]";
+        file_config << "y" << "[:";
+        for (size_t j = 0; j < transform.size(); j++) { file_config << transform[j].y; }
+        file_config << "]" << "}";
+    }
+    file_config << "]";
+    file_config.release();
+}
+
+#else
+
+typedef testing::TestWithParam< String > Objdetect_QRCode;
+TEST_P(Objdetect_QRCode, regression)
 {
     String root = cvtest::TS::ptr()->get_data_path() + "qrcode/";
-    // String cascades[] =
-    // {
-        // root + "haarcascade_frontalface_alt.xml",
-        // root + "lbpcascade_frontalface.xml",
-        // String()
-    // };
-
-    // vector<Rect> objects;
-    // RNG rng((uint64)-1);
-
-    // for( int i = 0; !cascades[i].empty(); i++ )
-    // {
-        // printf("%d. %s\n", i, cascades[i].c_str());
-        // CascadeClassifier cascade(cascades[i]);
-        // for( int j = 0; j < 100; j++ )
-        // {
-            // int width = rng.uniform(1, 100);
-            // int height = rng.uniform(1, 100);
-            // Mat img(height, width, CV_8U);
-            // randu(img, 0, 256);
-            // cascade.detectMultiScale(img, objects);
-        // }
-    // }
+    String dataset_config = cvtest::TS::ptr()->get_data_path() + "qrcode/dataset_config.json";
+    FileStorage file_config(dataset_config, FileStorage::READ);
+
+    if (file_config.isOpened())
+    {
+        String image_path = root + String(GetParam());
+        std::vector<Point> transform;
+        Mat src = imread(image_path, IMREAD_GRAYSCALE);
+        EXPECT_TRUE(detectQRCode(src, transform));
```

Comment:
What order of returned points? Is it well defined?

```.diff
-transform
+corners
```

Reply:
done, renamed "transform" and location points in a specific order